src/wallet/walletdb.cpp

```diff
@@ -1111,12 +1057,97 @@ static DBErrors LoadAddressBookRecords(CWallet* pwallet, DatabaseBatch& batch) E
+        DBErrors result = DBErrors::LOAD_OK;
+        uint256 hash;
+        key >> hash;
+        // LoadToWallet call below creates a new CWalletTx that fill_wtx
+        // callback fills with transaction metadata.
+        auto fill_wtx = [&](CWalletTx& wtx, bool new_tx) {
+            if(!new_tx) {
+                // There's some corruption here since the tx we just tried to load was already in the wallet.
+                // We don't consider this type of corruption critical, and can fix it by removing tx data and
+                // rescanning.
+                err = "Error: Corrupt transaction found. This can be fixed by removing transactions from wallet and rescanning.";
+                result = DBErrors::CORRUPT;
+                corrupted_tx = true;
+                return false;
+            }
+            value >> wtx;
+            if (wtx.GetHash() != hash)
+                return false;
+
+            // Undo serialize changes in 31600
+            if (31404 <= wtx.fTimeReceivedIsTxTime && wtx.fTimeReceivedIsTxTime <= 31703)
+            {
+                if (!value.empty())
+                {
+                    uint8_t fTmp;
+                    uint8_t fUnused;
+                    std::string unused_string;
+                    value >> fTmp >> fUnused >> unused_string;
+                    pwallet->WalletLogPrintf("LoadWallet() upgrading tx ver=%d %d %s\n",
+                                       wtx.fTimeReceivedIsTxTime, fTmp, hash.ToString());
+                    wtx.fTimeReceivedIsTxTime = fTmp;
+                }
+                else
+                {
+                    pwallet->WalletLogPrintf("LoadWallet() repairing tx ver=%d %s\n", wtx.fTimeReceivedIsTxTime, hash.ToString());
+                    wtx.fTimeReceivedIsTxTime = 0;
+                }
+                upgraded_txs.push_back(hash);
+            }
+
+            if (wtx.nOrderPos == -1)
+                any_unordered = true;
+
+            return true;
+        };
+        if (!pwallet->LoadToWallet(hash, fill_wtx)) {
+            if (corrupted_tx) {
+                result = DBErrors::CORRUPT;
+            } else {
+                result = DBErrors::NEED_RESCAN;
+            }
+        }
+        return result;
+    });
+    result = std::max(result, tx_res.m_result);
+
+    // Load locked utxo record
+    LoadResult locked_utxo_res = LoadRecords(pwallet, batch, DBKeys::LOCKED_UTXO,
+        [] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& err) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet) {
+        uint256 hash;
+        uint32_t n;
+        key >> hash;
+        key >> n;
+        pwallet->LockCoin(COutPoint(hash, n));
+        return DBErrors::LOAD_OK;
+    });
+    result = std::max(result, locked_utxo_res.m_result);
+
+    // Load orderposnext record
+    batch.Read(DBKeys::ORDERPOSNEXT, pwallet->nOrderPosNext);
```

Comment:
In commit "walletdb: refactor tx loading" (b26573ef95891bb69f00c3aaa67fb4678a82c818)

Again, there's two changes in behavior since this switching from `ReadKeyValue` to `Read`. One is that deserialization errors are ignored with no error or warning. The other is that this is switching from a prefix match to an exact match so keys that have other fields after the ORDERPOSNEXT string will be ignored.

I don't think either current or previous behavior is ideal, but I think I might opt to preserve previous behavior with:

```diff
diff --git a/src/wallet/walletdb.cpp b/src/wallet/walletdb.cpp
index 7292c6fda414..e6d86bf2ad30 100644
--- a/src/wallet/walletdb.cpp
+++ b/src/wallet/walletdb.cpp
@@ -1136,7 +1136,18 @@ static DBErrors LoadTxRecords(CWallet* pwallet, DatabaseBatch& batch, std::vecto
     result = std::max(result, locked_utxo_res.m_result);
 
     // Load orderposnext record
-    batch.Read(DBKeys::ORDERPOSNEXT, pwallet->nOrderPosNext);
+    LoadResult order_pos_res = LoadRecords(pwallet, batch, DBKeys::ORDERPOSNEXT,
+        [] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& err) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet) {
+        try {
+           value >> pwallet->nOrderPosNext;
+        } catch (const std::exception& e) {
+           if (err.empty()) {
+              err = e.what();
+           }
+           return DBErrors::NONCRITICAL_ERROR;
+        }
+        return DBErrors::LOAD_OK;
+    });
 
     return result;
 }

```

Reply:
Done